### PR TITLE
[Incremental Optimization] Combines https://github.com/apple/swift-driver/pull/557 with https://github.com/apple/swift-driver/pull/547

### PR DESCRIPTION
**Pull request by davidungar**
Combines better prior recovery, hash caching, and VirtualPath interning. *Very* preliminary measurement shows 14% less user time for a null-compilation with priors (with incremental imports), compared to the legacy driver. 11% less real time.

Marked DNM because the history needs to be cleaned up, or else this PR should be broken up into separate ones.

Thoughts?

EDIT: The history isn't as bad as I thought. @CodaFi , Do you want to merge the interning PR? If so, the 2nd commit in this one might help.

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
@swift-ci please test

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Node.swift [hunk 1]
```diff
@@ -54,12 +54,15 @@ extension ModuleDependencyGraph {
...
 
+    public let hashValue: Int
+
     /// This dependencySource is the file where the swiftDeps, etc. was read, not necessarily anything in the
     /// SourceFileDependencyGraph or the DependencyKeys
     init(key: DependencyKey, fingerprint: String?,
          dependencySource: DependencySource?) {
       self.keyAndFingerprint = try! KeyAndFingerprintHolder(key, fingerprint)
       self.dependencySource = dependencySource
+      self.hashValue = Self.computeHash(key, dependencySource)
```
Comment (cltnschlosser) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Node.swift:65]: This node is a class. Any advantage to making this lazily computed? Or are these always hashed and that just adds overhead?

Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Node.swift:65]: Good thought. I'll try it.

Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Node.swift:65]: Thanks for taking a look so soon!

Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Node.swift:65]: OK, I tried it. The lazy version is either slightly slower, or no faster, at least in the with-priors scenario. I suspect that's cause every node gets hashed into the `nodeFinder`. Good thought, though.

**Comment by davidungar** [comment]
I did some measurements (with this PR), using swift-driver itself as the test subject.
Looked at the elapsed time reported at the end of the build log.
I added an internal method to PrefixTrie, in an imported module, and rebuilt the the whole thing.
Old driver: ~7.9 secs. 
New Driver (with incremental imports) 4.2 secs.
New Driver (-disable-incremental-imports) 8.9 secs.

Can't wait to get this optimizations in!

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
Adding a comment from the closed PR that applies here:

Assuming I did everything correctly, I tried an invocation of the compiler when the priors were present. I used three versions of the new driver, plus the old driver. Version 1 of the new driver was main, version 2: main + your intern version (I had to rebase), version 3: the other version of the new driver was main + your intern version + the hash optimizations (just as in this PR). Times reported by the time command.

old driver: 0.95 (user), 1.05 (total)
new driver (main): 1.6 (user), 1.77 (total)
new driver (main + intern): 1.10 (user), 1.24 (total)
new driver (main + intern + hash opts): 0.81 (user), 1.07 (total)

It is certainly possible I did something wrong; would love you to try the experiment for yourself. Contact me in Slack for details.

**Comment by CodaFi** [comment]
These PRs appear to have been merged. Do you still need this?

**Comment by davidungar** [comment]
> These PRs appear to have been merged. Do you still need this?

No, tx for the ping.

**Pull request closed without merging**